### PR TITLE
Added validation and pulled Countries/Cities from server for the "General Info" step #135

### LINKED ISSUE #135: (SP: 0.5) Add validation for General info step
UserStory: #58 , #59 
Blocked by: #132 

## Description
Add validation to required fields in General Info step for both student and tutor steppers

## Design 

Image

## Acceptance Criteria
- [x] First name and Last name fields must be required and show an error when user removed text from them.
- [x] Country and city have to be pulled from server on input focus
- [x] Length of additional info can't be more than 100 symbols and block texting when it reached limit
- [ ] When there are errors in required fields stepper can't be finished **[ Will be addressed in issue #164 ]**

### PR DESCRIPTION
**Pull request by Olenka-Hryk**
### Added validation to required fields and pulled Countries/Cities from the server for the "General Info" step (for both `student` and `tutor` steppers) [ Close #135 ]

- [x] `First name` and `Last name` fields are required and show an error when the user removes text from them:
-When the user clears the fields `First name` and `Last name`, an error is displayed: **"This field cannot be empty"**
-When the user enters invalid data into the`First name` and `Last name`, an error is displayed: **"This field cannot be shorter than 2 and longer than 15 characters"**
- [x] Length of `additional info` can't exceed 100 characters. Text input is blocked when the limit is reached, and an error message is displayed: **"This text is too long"**:
Screenshot 2024-05-13 at 20 39 40

- [x] Country and city data are fetched from the server when the input field is focused

| Fetching Countries                                      | Fetching Cities                                      |
|---------------------------------------------------------|------------------------------------------------------|
| Fetching Countries | Fetching Cities |

- [ ] When there are errors in required fields stepper can't be finished **[ Will be addressed in issue #164 ]**

## Important Changes: 
- [Updated localization dictionary for "GeneralInfoStep"](https://github.com/Space2Study-UA-1156/Client-02/commit/78b683d3a60e90a29d5db8d76949a85422c9aa71)
- [Implemented useMemo optimization in the stepper context file](https://github.com/Space2Study-UA-1156/Client-02/commit/6b2b829dd840447b19f3970b40ae97359fe6174a)
- [Added method "handleSelectChange" to useForm hook & optimization](https://github.com/Space2Study-UA-1156/Client-02/commit/2936e8154dc4d7adc057e2f457353a14e6a4e479)
- [Added useForm hook to StepContext](https://github.com/Space2Study-UA-1156/Client-02/commit/1af5322df6b0fa035cb9ab9468af779ce46366a6)
- **Added REST Cache for Axios Client**:
_Why it was added:_ It was added to enhance the performance and efficiency of data retrieval by caching responses from the server. This reduces the need for repeated network requests, resulting in faster data access

| **BEFORE cache:**                                              | **AFTER cache:**                                               |
|----------------------------------------------------------------|----------------------------------------------------------------|
| BEFORE cache | AFTER cache |

##  Breaking Changes:
A new package, "`axios-cache-interceptor`" version "`^0.10.3`", has been added.
After merging this PR, you will need to pull new data to your `develop` branch and **run the following command**:
`npm run reinstall`

### CODE REVIEW

**Review by Nazar-Lysak — APPROVED** [review]

**Review by docia — APPROVED** [review]

**Review by BodiaPutiak — APPROVED** [review]

File: src/containers/tutor-home-page/general-info-step/GeneralInfoStep.jsx [hunk 1]
```diff
@@ -60,63 +66,70 @@ const GeneralInfoStep = ({ btnsBox }) => {
...
           placeholder={t('becomeTutor.generalInfo.textFieldLabel')}
           rows={4}
-          value={stepData?.generalInfo?.data?.professionalSummary ?? ''}
+          type='text'
+          value={data.professionalSummary}
         />
         <Typography
           sx={styles.summaryLength}
-        >{`${stepData?.generalInfo?.data?.professionalSummary?.length}/70`}</Typography>
+        >{`${data.professionalSummary?.length}/100`}</Typography>
```
Comment (vlad-khrychov) [inline src/containers/tutor-home-page/general-info-step/GeneralInfoStep.jsx]: I think `100` should be saved in some constant for using the same value (or +1) on 111 line.

File: src/context/step-context.jsx [hunk 2]
```diff
@@ -1,40 +1,61 @@
...
+  })
+
+  useEffect(() => {
+    setNeedConfirmation(isDirty)
+  }, [isDirty, setNeedConfirmation])
+
   const [generalData, setGeneralData] = useState({
     data: initialValues,
-    errors: {}
+    errors: errors
```
Comment (vlad-khrychov) [inline src/context/step-context.jsx]:
```suggestion
    errors
```

**Review by vlad-khrychov — APPROVED** [review]

File: src/context/step-context.jsx (hunk 2 above)
Comment (Olenka-Hryk) [reply src/context/step-context.jsx]: Changed it!

File: src/containers/tutor-home-page/general-info-step/GeneralInfoStep.jsx (hunk 1 above)
Comment (Olenka-Hryk) [reply src/containers/tutor-home-page/general-info-step/GeneralInfoStep.jsx]: Thank you for the advice! I created a `maxLengthTextField` field in the `constants.js` file and imported this value into `GeneralInfoStep.jsx`.

**Review by vlad-khrychov — APPROVED** [review]

**Pull request merged**